Can't invoke Streamable HTTP MCP server implemented with Micronaut MCP

I implemented a Streamable HTTP MCP server with the Micronaut MCP module:
https://github.com/glaforge/mn-mcp-server

It exposes a couple methods:
* `current-moon-phase()`: to get the current phase of the moon
* `moon-phase-at-date("2025-09-28")`: to get the phase at a certain date

You can call that server here with the MCP Inspector for testing:
https://mn-mcp-server-1029513523185.europe-west1.run.app/mcp/
The MCP inspector is totally fine with it, no error, I can list all the tools, call all the methods, etc.

I tried with the LangChain4j MCP module as well (which is a reimplementation of the protocol), with the following code:
https://gist.github.com/glaforge/809ccbad223886b0db03e679adcb7e85
And it also works totally fine with my server.

However, when I can from the Agent Development Kit for Java (ADK), it fails.
As ADK is using the official MCP SDK, I tried to reproduce the problem with the SDK directly, to remove ADK from the equation, and it fails in the same way.

I reproduced the problem with the following code:
https://gist.github.com/glaforge/741edcad883b27268ab10249400ce6d6

I'm seeing this stacktrace (same as within ADK for Java):
https://gist.github.com/glaforge/afd1194ece6bc909b091a946c030dc27

i.e.
```
io.modelcontextprotocol.spec.McpTransportException: Invalid SSE response. Status code: 405 Line: {"message":"Method Not Allowed","_links":{"self":{"href":"/mcp","templated":false}},"_embedded":{"errors":[{"message":"Method [GET] not allowed for URI [/mcp]. Allowed methods: [POST]"}]}}
	at io.modelcontextprotocol.client.transport.ResponseSubscribers$SseLineSubscriber.hookOnNext(ResponseSubscribers.java:185)
	at io.modelcontextprotocol.client.transport.ResponseSubscribers$SseLineSubscriber.hookOnNext(ResponseSubscribers.java:72)
...
21:04:44.389 [HttpClient-1-Worker-0] WARN io.modelcontextprotocol.client.transport.HttpClientStreamableHttpTransport -- Unknown media type text/html returned for POST in session [missing_session_id]
21:04:44.389 [HttpClient-1-Worker-0] WARN io.modelcontextprotocol.client.LifecycleInitializer -- Handling exception
java.lang.RuntimeException: Unknown media type returned: text/html
	at io.modelcontextprotocol.client.transport.HttpClientStreamableHttpTransport.lambda$sendMessage$27(HttpClientStreamableHttpTransport.java:522)
...
Caused by: java.lang.RuntimeException: Unknown media type returned: text/html
	at io.modelcontextprotocol.client.transport.HttpClientStreamableHttpTransport.lambda$sendMessage$27(HttpClientStreamableHttpTransport.java:522)
```

So I'm not sure if it's a problem on Micronaut's side or on the MCP SDK's side.
Maybe one is too strict with the spec, or not tolerant enough, I don't know.

But when I read the spec:
```
The server MUST provide a single HTTP endpoint path (hereafter referred to as the MCP endpoint) 
that supports both POST and GET methods. For example, this could be a URL like 
https://example.com/mcp.
```
```
The client MAY issue an HTTP GET to the MCP endpoint. This can be used to open an SSE stream, 
allowing the server to communicate to the client, without the client first sending data via HTTP POST.
```
```
The server MUST either return Content-Type: text/event-stream in response to this HTTP GET, 
or else return HTTP 405 Method Not Allowed, indicating that the server does not offer an SSE 
stream at this endpoint.
```
It seems that Micronaut does the right thing: if a client tries to do a GET to do an SSE connection, 
it gets a 405, which is the error Micronaut is returning.

Any idea what could be going wrong here @tzolov ?

/cc @sdelamo @graemerocher @Poggecci

--- a/mcp-core/src/main/java/io/modelcontextprotocol/client/transport/HttpClientSseClientTransport.java
+++ b/mcp-core/src/main/java/io/modelcontextprotocol/client/transport/HttpClientSseClientTransport.java
@@ -18,14 +18,13 @@ import java.util.function.Function;
 
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
-
-import io.modelcontextprotocol.json.McpJsonMapper;
-import io.modelcontextprotocol.json.TypeRef;
-
+import io.modelcontextprotocol.client.transport.ResponseSubscribers.ResponseEvent;
 import io.modelcontextprotocol.client.transport.customizer.McpAsyncHttpClientRequestCustomizer;
 import io.modelcontextprotocol.client.transport.customizer.McpSyncHttpClientRequestCustomizer;
-import io.modelcontextprotocol.client.transport.ResponseSubscribers.ResponseEvent;
 import io.modelcontextprotocol.common.McpTransportContext;
+import io.modelcontextprotocol.json.McpJsonMapper;
+import io.modelcontextprotocol.json.TypeRef;
+import io.modelcontextprotocol.spec.HttpHeaders;
 import io.modelcontextprotocol.spec.McpClientTransport;
 import io.modelcontextprotocol.spec.McpSchema;
 import io.modelcontextprotocol.spec.McpSchema.JSONRPCMessage;
@@ -469,7 +468,7 @@ public class HttpClientSseClientTransport implements McpClientTransport {
 		return Mono.deferContextual(ctx -> {
 			var builder = this.requestBuilder.copy()
 				.uri(requestUri)
-				.header("Content-Type", "application/json")
+				.header(HttpHeaders.CONTENT_TYPE, "application/json")
 				.header(MCP_PROTOCOL_VERSION_HEADER_NAME, MCP_PROTOCOL_VERSION)
 				.POST(HttpRequest.BodyPublishers.ofString(body));
 			var transportContext = ctx.getOrDefault(McpTransportContext.KEY, McpTransportContext.EMPTY);

--- a/mcp-core/src/main/java/io/modelcontextprotocol/client/transport/HttpClientStreamableHttpTransport.java
+++ b/mcp-core/src/main/java/io/modelcontextprotocol/client/transport/HttpClientStreamableHttpTransport.java
@@ -246,7 +246,7 @@ public class HttpClientStreamableHttpTransport implements McpClientTransport {
 				}
 
 				var builder = requestBuilder.uri(uri)
-					.header("Accept", TEXT_EVENT_STREAM)
+					.header(HttpHeaders.ACCEPT, TEXT_EVENT_STREAM)
 					.header("Cache-Control", "no-cache")
 					.header(HttpHeaders.PROTOCOL_VERSION, MCP_PROTOCOL_VERSION)
 					.GET();
@@ -371,7 +371,7 @@ public class HttpClientStreamableHttpTransport implements McpClientTransport {
 
 		BodyHandler<Void> responseBodyHandler = responseInfo -> {
 
-			String contentType = responseInfo.headers().firstValue("Content-Type").orElse("").toLowerCase();
+			String contentType = responseInfo.headers().firstValue(HttpHeaders.CONTENT_TYPE).orElse("").toLowerCase();
 
 			if (contentType.contains(TEXT_EVENT_STREAM)) {
 				// For SSE streams, use line subscriber that returns Void
@@ -420,9 +420,9 @@ public class HttpClientStreamableHttpTransport implements McpClientTransport {
 				}
 
 				var builder = requestBuilder.uri(uri)
-					.header("Accept", APPLICATION_JSON + ", " + TEXT_EVENT_STREAM)
-					.header("Content-Type", APPLICATION_JSON)
-					.header("Cache-Control", "no-cache")
+					.header(HttpHeaders.ACCEPT, APPLICATION_JSON + ", " + TEXT_EVENT_STREAM)
+					.header(HttpHeaders.CONTENT_TYPE, APPLICATION_JSON)
+					.header(HttpHeaders.CACHE_CONTROL, "no-cache")
 					.header(HttpHeaders.PROTOCOL_VERSION, MCP_PROTOCOL_VERSION)
 					.POST(HttpRequest.BodyPublishers.ofString(jsonBody));
 				var transportContext = ctx.getOrDefault(McpTransportContext.KEY, McpTransportContext.EMPTY);
@@ -459,15 +459,19 @@ public class HttpClientStreamableHttpTransport implements McpClientTransport {
 
 					String contentType = responseEvent.responseInfo()
 						.headers()
-						.firstValue("Content-Type")
+						.firstValue(HttpHeaders.CONTENT_TYPE)
 						.orElse("")
 						.toLowerCase();
 
-					if (contentType.isBlank()) {
-						logger.debug("No content type returned for POST in session {}", sessionRepresentation);
+					String contentLength = responseEvent.responseInfo()
+						.headers()
+						.firstValue(HttpHeaders.CONTENT_LENGTH)
+						.orElse(null);
+
+					if (contentType.isBlank() || "0".equals(contentLength)) {
+						logger.debug("No body returned for POST in session {}", sessionRepresentation);
 						// No content type means no response body, so we can just
-						// return
-						// an empty stream
+						// return an empty stream
 						deliveredSink.success();
 						return Flux.empty();
 					}

--- a/mcp-core/src/main/java/io/modelcontextprotocol/spec/HttpHeaders.java
+++ b/mcp-core/src/main/java/io/modelcontextprotocol/spec/HttpHeaders.java
@@ -26,4 +26,31 @@ public interface HttpHeaders {
 	 */
 	String PROTOCOL_VERSION = "MCP-Protocol-Version";
 
+	/**
+	 * The HTTP Content-Length header.
+	 * @see <a href=
+	 * "https://httpwg.org/specs/rfc9110.html#field.content-length">RFC9110</a>
+	 */
+	String CONTENT_LENGTH = "Content-Length";
+
+	/**
+	 * The HTTP Content-Type header.
+	 * @see <a href=
+	 * "https://httpwg.org/specs/rfc9110.html#field.content-type">RFC9110</a>
+	 */
+	String CONTENT_TYPE = "Content-Type";
+
+	/**
+	 * The HTTP Accept header.
+	 * @see <a href= "https://httpwg.org/specs/rfc9110.html#field.accept">RFC9110</a>
+	 */
+	String ACCEPT = "Accept";
+
+	/**
+	 * The HTTP Cache-Control header.
+	 * @see <a href=
+	 * "https://httpwg.org/specs/rfc9111.html#field.cache-control">RFC9111</a>
+	 */
+	String CACHE_CONTROL = "Cache-Control";
+
 }

--- a/mcp-spring/mcp-spring-webflux/src/main/java/io/modelcontextprotocol/client/transport/WebClientStreamableHttpTransport.java
+++ b/mcp-spring/mcp-spring-webflux/src/main/java/io/modelcontextprotocol/client/transport/WebClientStreamableHttpTransport.java
@@ -293,9 +293,10 @@ public class WebClientStreamableHttpTransport implements McpClientTransport {
 					// 200 OK for notifications
 					if (response.statusCode().is2xxSuccessful()) {
 						Optional<MediaType> contentType = response.headers().contentType();
+						long contentLength = response.headers().contentLength().orElse(-1);
 						// Existing SDKs consume notifications with no response body nor
 						// content type
-						if (contentType.isEmpty()) {
+						if (contentType.isEmpty() || contentLength == 0) {
 							logger.trace("Message was successfully sent via POST for session {}",
 									sessionRepresentation);
 							// signal the caller that the message was successfully

